beacon/light/committee_chain.go

```diff
@@ -0,0 +1,555 @@
+	}
+	s.updates.add(batch, period, update)
+	if err := batch.Write(); err != nil {
+		log.Error("Error writing batch into chain database", "error", err)
+		return err
+	}
+	log.Info("Inserted new committee update", "period", period, "next committee root", update.NextSyncCommitteeRoot)
+	return nil
+}
+
+// NextSyncPeriod returns the next period where an update can be added and also
+// whether the chain is initialized at all.
+func (s *CommitteeChain) NextSyncPeriod() (uint64, bool) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	if s.committees.IsEmpty() {
+		return 0, false
+	}
+	if !s.updates.IsEmpty() {
+		return s.updates.AfterLast, true
+	}
+	return s.committees.AfterLast - 1, true
+}
+
+// rollback removes all committees and fixed roots from the given period and updates
+// starting from the previous period.
+func (s *CommitteeChain) rollback(batch ethdb.Batch, period uint64) {
+	s.deleteCommitteesFrom(batch, period)
+	s.fixedRoots.deleteFrom(batch, period)
+	if period > 0 {
+		period--
+	}
+	s.updates.deleteFrom(batch, period)
+}
+
+// getCommitteeRoot returns the committee root at the given period, either fixed,
+// proven by a previous update or both. It returns an empty hash if the committee
+// root is unknown.
+func (s *CommitteeChain) getCommitteeRoot(period uint64) common.Hash {
+	if root := s.fixedRoots.get(period); root != (common.Hash{}) || period == 0 {
+		return root
+	}
+	if update := s.updates.get(period - 1); update != nil {
+		return update.NextSyncCommitteeRoot
+	}
+	return common.Hash{}
+}
+
+// getSyncCommittee returns the deserialized sync committee at the given period.
+func (s *CommitteeChain) getSyncCommittee(period uint64) syncCommittee {
+	if c, ok := s.syncCommitteeCache.Get(period); ok {
+		return c
+	}
+	if sc := s.committees.get(period); sc != nil {
+		c, err := s.sigVerifier.deserializeSyncCommittee(sc)
+		if err != nil {
+			log.Error("Sync committee deserialization error", "error", err)
+			return nil
+		}
+		s.syncCommitteeCache.Add(period, c)
+		return c
+	}
+	log.Error("Missing serialized sync committee", "period", period)
+	return nil
+}
+
+// VerifySignedHeader returns true if the given signed header has a valid signature
+// according to the local committee chain. The caller should ensure that the
+// committees advertised by the same source where the signed header came from are
+// synced before verifying the signature.
+// The age of the header is also returned (the time elapsed since the beginning
+// of the given slot, according to the local system clock). If enforceTime is
+// true then negative age (future) headers are rejected.
+func (s *CommitteeChain) VerifySignedHeader(head types.SignedHeader) (bool, time.Duration) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	return s.verifySignedHeader(head)
+}
+
+func (s *CommitteeChain) verifySignedHeader(head types.SignedHeader) (bool, time.Duration) {
+	var (
+		slotTime = int64(time.Second) * int64(s.config.GenesisTime+head.Header.Slot*12)
```

Comment:
`s.config.GenesisTime+head.Header.Slot*12` might overflow `int64` (on malicious input)

Reply:
Added a check, now it should be safe.